### ISSUE TITLE
SPARQL Front end

### ISSUE THREAD

**Issue by joshuaauerbachwatson** [querycert/qcert#5]
The SPARQL language is designed to formulate queries on RDF graphs stores and it has properties that are similar to rule languages and relational languages.  I think it might be interesting to produce a SPARQL front-end for qcert that works by translating SPARQL queries to Rules and CAMP.

I would use Apache Jena's ARQ component to manage the parsing, initial analysis and visiting of SPARQL, producing the planned Java ASTs for Rules and CAMP as described in issue #2.

Here's what I am _not_ saying at this point.
- I'm _not_ (yet) saying that translating SPARQL to CAMP makes semantic sense (I speculate that it does, but it is unclear to what extent the RDF graph data model is fundamentally at odds with the nested relational model with aggregation assumed by qcert)
- I'm not saying that translating SPARQL to any of targets that qcert supports is of any interest to anybody (I used to have to worry about things like this when I did software for a living, but now I do it just for fun so I can suspend worrying about possible impact for a lot longer)

Hopefully the answers to both will become clear down the road.

**Comment by joshuaauerbachwatson** [comment]
Now that the Java AST classes are accepted into the master repository, it is time to start on actual SPARQL translation.   Work is resuming in the fork joshuaauerbachwatson/qcert and will result in another pull request when there is something to show.

Initially, there is just the Java project sparql2camp which has the means to manage the dependency on Jena ARQ and on the SPARQL conformance tests in https://www.w3.org/2001/sw/DataAccess/r2.   No actual code yet but that's next.

The plan is to translate all SPARQL non-mutating operations on the triple store; there are no plans to translate update operations (there are defined in the current version of SPARQL but I'm considering them out of scope).   There will undoubtedly be other omissions but I don't know what they are yet.

I think I can answer one initial question (does the translation make semantic sense) in a limited, equivocal sense:  yes, it does, for all the good it does us.  A triple store can certainly be viewed as a relation.   The literature discusses many representations, both normalized and denormalized, but this is all about how many copies of things exist and in what places, not the basic idea of what a triple is, which remains as an unchanging basis of all the work.   Basically, the semantics are defined as operations on a sea of triples; the "graph" is implicit in the fact that many triple objects are references, so the graph is reconstructed on the fly by doing lots of joins.   Lots and lots of joins.  By denormalizing the data and playing around with the representation you can reduce the number of joins.   But, there are still going to be plenty of them.

Now, most of the literature seems to be about efficiency; no one really questions the semantics, so, in that sense I we are "ok" and I think the key question is this.

Translating SPARQL into SQL (which is what is mostly discussed in the few papers I've looked at) is usually done knowing the details of how the data is going to be shredded.   Since CAMP and NRA are relational languages, we might not escape the need for this.  But, I'd like to start with the possibly absurd premise that yes we can escape the need.   That is, I'm going to use a naive data model in which the operations on working memory assume that memory is just filled with triples, no duplication, nothing fancy.    So, the CAMP translation will make heavy use of the "fetch macro" (basically a specialized join) and of more general joins where it makes sense.  This has the "nice" property that you can delay further optimizations until you know the backend target.  It has the "nasty" property that, for many and maybe all backends, the translation will be pessimal and very hard to optimize.   We'll see.

**Comment by joshuaauerbachwatson** [comment]
When I recently merged and abandoned my fork of the qcert repository, I declined to merge the limited work I had done on SPARQL, so it is now orphaned and continually rotting.  I think it makes sense to clear out this issue.  If SPARQL ever again becomes interesting, it would not be hard to start over, perhaps in OCaml rather than Java (although the existence of Jena ARQ makes Java attractive).  Note that, if Java is again chosen, the work to create Java ASTs for Rules and CAMP is among what was merged and is present in this repository.
